haystack/lib/cli.ex

```diff
@@ -0,0 +1,145 @@
+defmodule CLI do
+  @moduledoc """
+  CLI is a module for interacting with Haystack through the commandline. 
+  [ Usage ]
+    Fragmentation
+    fragment --in <string> --count <integer> --keyfile <string> --out <string>
+    fragment --in <string> --count <integer> --out <string>
+    
+    Reassembly
+    reassemble --in <string> --keyfile <path/to/keyfile> --out <string>
+    reassemble --in <string> --out <string>
+    
+  [ Parameters ]
+    --in | -i
+      path to the input file (for fragmentation) or the input directory (for reassembly).
+    --count | -c
+      number of fragments to split the file into during fragmentation.
+    --keyfile | -k
+      path to a file containing the plaintext password to be used for encryption (for fragmentation) or decryption(for reassembly).
+    --out | -o
+      path to the output directory where fragments (for fragmentation) or the original file (for reassembly) will be located.
+  """
+
+  def main(argv) do
+    parse_args(argv)
+  end
+
+  defp parse_args(argv) do
+    switches = [
+      in: :string, 
+      count: :integer, 
+      keyfile: :string, 
+      out: :string, 
+      save_orig: :boolean,
+    ]
+    aliases = [
+      i: :in, 
+      c: :count, 
+      k: :keyfile, 
+      o: :out, 
+      s: :save_orig,
+    ]
+    parse = OptionParser.parse(argv, switches: switches, aliases: aliases)
+    case parse do
+      {opts,  ["fragment"], _}   -> map_params(opts, :fragment)
+      {opts,  ["reassemble"], _} -> map_params(opts, :reassemble)
+      {_opts, ["help" | _], _}   -> help_dialog()
+      _ -> IO.puts "Invalid parameters"
+    end
+  end
+
+  # Maps given parameters to the correct handler for the given 
+  # password entry method.
+  defp map_params(opts, :fragment) do
+    # TODO validate that all necessary params are not nil before calling function!
+    case opts[:keyfile] do
```

Comment:
I think this is one of those cases where removing the conditional would trigger better code:

~~~ elixir
defp map_params(opts, :fragment) do
Haystack.fragment(
        opts[:in],
        opts[:count],
        opts[:keyfile] |> maybe_get_password(),
        opts[:out],
        opts[:save_orig]
      )
end

def maybe_get_password(nil) do
  password_get("Enter the password to encrypt with: ")
end

def maybe_get_password(keyfile) do
  Utils.File.parse_keyfile(keyfile)
end